### DISCUSSION TITLE
What is the difference between 'flag' and true of optimization.sideEffects

### DISCUSSION THREAD

**Discussion by galaxy-s10** [webpack/webpack#16315] [Q&A]
```js
/**
 * Skip over modules which contain no side effects when exports are not used (false: disabled, 'flag': only use manually placed side effects flag, true: also analyse source code for side effects).
 */
sideEffects?: "flag" | boolean;
```
i don't understand "flag" and true，In my opinion, these two are the same。

**Comment by galaxy-s10** [comment]
hi @alexander-akait ,can you explain this？

**Comment by galaxy-s10** [comment]
@sokra ,hi i read https://github.com/webpack/webpack/blob/main/examples/side-effects/README.md,but it no say 'flag' and true difference, you know it or no

**Comment by galaxy-s10** [comment]
I tested the two. No matter 'flag' or 'true', they will not analyze the side effects of the source code. Even if a module has side effects, they will both delete it. But the official document says that when true, they will also analyze the source code, but in fact, it seems that there is no analysis of the source code. The behavior of true is the same as that of 'flag'

**Comment by galaxy-s10** [comment]
this is my demo: https://github.com/galaxy-s10/issue-demo/blob/master/demo1/webpack.config.js, When package.json is false, regardless of optimization Whether the sideEffects value is' flag 'or true, the final code results are the same, and the side effects of the source code will not be checked

**Comment by alexander-akait** [comment]
Sorry, it is not personal chat for help, please be patient in future and wait the answer

**Comment by alexander-akait** [comment]
`sideEffects: 'flag'` 

> To only use the manual flag and do not analyse source code:
> The 'flag' value is used by default in non-production builds.

With `'flag'` webpack skips analyze code (i.e. only `package.json` flags used), so it is faster, you need to use `true` in the `production` node

**Reply by galaxy-s10** [reply]
However, if you set true in the production mode, a library is used In json, sideEffects: false is used. The final result will still be based on the sideEffects declared by the library, and the side effects of the library will be deleted. I want to know whether this is the default behavior of the webpack or a bug of the webpack. Optimization is reasonable After sideEffects is set to true, I understand whether webpack will analyze the source code of the library and output the side effect code to the bundle, instead of still following the sideEffects field of package.json of the library

**Reply by alexander-akait** [reply]
Not only sideEffects can detected not used things, so removing something can not only based on `sideEffects`, wihtout real code and example I can't say, sorry

**Reply by galaxy-s10** [reply]
Recurrence demo: https://github.com/galaxy-s10/issue-demo/tree/master/demo2
first git clone it ，yarn install
and then, move aaa into node_modules
last npm run build，when optimization.sideEffects is true and mode is production, still use aaa package.json's sideEffects，but should Subject to the configuration of webpack

**Reply by galaxy-s10** [reply]
i feel this is a bug ，or webpack Intentionally

**Reply by alexander-akait** [reply]
I don't think it is the bug, all works as expected, 

When you have `"sideEffects": true` our output:
```
(()=>{"use strict";console.log("hi")})();
``` 

i.e. we keep module because it has side effect

But if you set `"sideEffects": false` our output is empty, because webpack knows - no side effects and can drop the whole module (and nested modules)

**Reply by galaxy-s10** [reply]
> 我不认为这是错误，一切都按预期工作，
> 
> 当你有`"sideEffects": true`我们的输出时：
> 
> ```
> (()=>{"use strict";console.log("hi")})();
> ```
> 
> 即我们保留模块，因为它有副作用
> 
> 但是如果你设置`"sideEffects": false`我们的输出为空，因为 webpack 知道 - 没有副作用并且可以删除整个模块（和嵌套模块）

I think you misunderstood me，i mean change webpack config's optimization.sideEffects is true or 'flag', In any case, the value of true or flag does not affect the output

**Reply by galaxy-s10** [reply]
webpack config's optimization.sideEffects is true or 'flag'，the output both empty

**Reply by alexander-akait** [reply]
We do this when you have `true` - https://github.com/webpack/webpack/blob/main/lib/optimize/SideEffectsFlagPlugin.js#L114, as you can see we look at code and search `pure` comments

**Reply by galaxy-s10** [reply]
> 当您拥有时，我们会这样做`true`- https://github.com/webpack/webpack/blob/main/lib/optimize/SideEffectsFlagPlugin.js#L114，正如您所看到的，我们查看代码并搜索`pure`评论

i look a look
